File: backend/src/main/java/com/expensetracker/controller/ExpenseController.java
package com.expensetracker.controller;

import com.expensetracker.model.Expense;
import com.expensetracker.service.ExpenseService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDate;
import java.util.List;

@RestController
@RequestMapping("/api/expenses")
@CrossOrigin(origins = "*")
public class ExpenseController {

    @Autowired
    private ExpenseService expenseService;

    @GetMapping
    public ResponseEntity<List<Expense>> getAllExpenses() {
        return ResponseEntity.ok(expenseService.getAllExpenses());
    }

    @GetMapping("/{id}")
    public ResponseEntity<Expense> getExpenseById(@PathVariable Long id) {
        return expenseService.getExpenseById(id)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @PostMapping
    public ResponseEntity<Expense> createExpense(@Valid @RequestBody Expense expense) {
        Expense createdExpense = expenseService.createExpense(expense);
        return ResponseEntity.status(HttpStatus.CREATED).body(createdExpense);
    }

    @PutMapping("/{id}")
    public ResponseEntity<Expense> updateExpense(@PathVariable Long id, @Valid @RequestBody Expense expense) {
        try {
            Expense updatedExpense = expenseService.updateExpense(id, expense);
            return ResponseEntity.ok(updatedExpense);
        } catch (RuntimeException e) {
            return ResponseEntity.notFound().build();
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteExpense(@PathVariable Long id) {
        expenseService.deleteExpense(id);
        return ResponseEntity.noContent().build();
    }

    @GetMapping("/category/{category}")
    public ResponseEntity<List<Expense>> getExpensesByCategory(@PathVariable String category) {
        return ResponseEntity.ok(expenseService.getExpensesByCategory(category));
    }

    @GetMapping("/date-range")
    public ResponseEntity<List<Expense>> getExpensesByDateRange(
            @RequestParam @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate startDate,
            @RequestParam @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate endDate) {
        return ResponseEntity.ok(expenseService.getExpensesByDateRange(startDate, endDate));
    }

    @GetMapping("/categories")
    public ResponseEntity<List<String>> getAllCategories() {
        return ResponseEntity.ok(expenseService.getAllCategories());
    }
}

File: backend/src/main/kotlin/com/expensetracker/controller/ExpenseController.kt
package com.expensetracker.controller

import com.expensetracker.model.Expense
import com.expensetracker.service.ExpenseService
import jakarta.validation.Valid
import org.springframework.format.annotation.DateTimeFormat
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.time.LocalDate

@RestController
@RequestMapping("/api/expenses")
@CrossOrigin(origins = ["*"])
class ExpenseController(
    private val expenseService: ExpenseService
) {

    @GetMapping
    fun getAllExpenses(): ResponseEntity<List<Expense>> {
        return ResponseEntity.ok(expenseService.getAllExpenses())
    }

    @GetMapping("/{id}")
    fun getExpenseById(@PathVariable id: Long): ResponseEntity<Expense> {
        return expenseService.getExpenseById(id)
            ?.let { ResponseEntity.ok(it) }
            ?: ResponseEntity.notFound().build()
    }

    @PostMapping
    fun createExpense(@Valid @RequestBody expense: Expense): ResponseEntity<Expense> {
        val createdExpense = expenseService.createExpense(expense)
        return ResponseEntity.status(HttpStatus.CREATED).body(createdExpense)
    }

    @PutMapping("/{id}")
    fun updateExpense(
        @PathVariable id: Long,
        @Valid @RequestBody expense: Expense
    ): ResponseEntity<Expense> {
        return try {
            val updatedExpense = expenseService.updateExpense(id, expense)
            ResponseEntity.ok(updatedExpense)
        } catch (e: RuntimeException) {
            ResponseEntity.notFound().build()
        }
    }

    @DeleteMapping("/{id}")
    fun deleteExpense(@PathVariable id: Long): ResponseEntity<Void> {
        expenseService.deleteExpense(id)
        return ResponseEntity.noContent().build()
    }

    @GetMapping("/category/{category}")
    fun getExpensesByCategory(@PathVariable category: String): ResponseEntity<List<Expense>> {
        return ResponseEntity.ok(expenseService.getExpensesByCategory(category))
    }

    @GetMapping("/date-range")
    fun getExpensesByDateRange(
        @RequestParam @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) startDate: LocalDate,
        @RequestParam @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) endDate: LocalDate
    ): ResponseEntity<List<Expense>> {
        return ResponseEntity.ok(expenseService.getExpensesByDateRange(startDate, endDate))
    }

    @GetMapping("/categories")
    fun getAllCategories(): ResponseEntity<List<String>> {
        return ResponseEntity.ok(expenseService.getAllCategories())
    }
}
